python/freetensor/core/transformer.py

```diff
@@ -971,26 +991,69 @@ def into_staging(func, caller_env, src: str = None, verbose=False):
         lineno = 1
         file = f'<staging:{func.__name__}>'
 
-    tree = ast.parse(process_annotating_comments(src))
-    tree = ast.fix_missing_locations(Transformer(file, lineno).visit(tree))
+    func.__globals__['__freetensor__'] = sys.modules['freetensor']
+    func_locals = extra_locals
+    if func.__closure__:
+        func_locals.update({
+            name: cell.cell_contents
+            for name, cell in zip(func.__code__.co_freevars, func.__closure__)
+        })
 
-    import astor
-    source = astor.to_source(tree)
+    tree = ast.parse(process_annotating_comments(src))
+    tree = Transformer(file, lineno).visit(tree)
+
+    # Wrap the staging function.
+    # This is to workaround an issue of CPython.
+    # (See https://github.com/python/cpython/issues/86084)
+    # Without a context, CPython has no idea whether a name is from globals or
+    # locals, thus it uses LOAD_GLOBAL for variables from the closure.
+    # To avoid this, we wrap the staging function with the closure variables
+    # passed as arguments to the wrapper.
+    WRAPPER_NAME = '__freetensor_staging_wrapper__'
+    tree.body = [
+        ast.FunctionDef(
+            name=WRAPPER_NAME,
+            args=ast.arguments(
+                posonlyargs=[],
+                args=[ast.arg(name, None) for name in func_locals.keys()],
+                vararg=None,
+                kwonlyargs=[],
+                kw_defaults=[],
+                kwarg=None,
+                defaults=[]),
+            body=[
+                *tree.body,
+                ast.Return(value=ast.Name(id=func.__name__, ctx=ast.Load()))
+            ],
+            decorator_list=[],
+            returns=None),
+    ]
+    tree = ast.fix_missing_locations(tree)
 
     if verbose:
+        import astor
+        source = astor.to_source(tree)
+
         from pygments import highlight
         from pygments.lexers import PythonLexer
         from pygments.formatters import TerminalFormatter
         print(highlight(source, PythonLexer(),
                         TerminalFormatter(bg='dark', linenos=True)),
               file=sys.stderr)
 
-    caller_env['freetensor'] = sys.modules['freetensor']
-    exec(compile(source, f'<staging:{func.__name__}>', 'exec'), caller_env)
```

Comment:
I see originally everything in `caller_env` were treated as globals, but now they are locals. Although variables in `caller_env` are truly locals, but what was the problem when we treated them as locals?

Reply:
Actually the problem is not fully resolved by this PR. Current implementation only passes the globals as-is, but not the closure. If `foo` and `bar` are defined inside some function, the example will not work since the `cell_content` has nothing yet, and we cannot fill `func_locals` properly. I will try to figure out some workaround on this.

Reply:
> For example, it would be able to do below after this change:
> 
> ```python
> @transform
> def foo():
>     bar()
> 
> def bar():
>     ...
> ```
> 
> Which is allowed in Python but not allowed by the previous transformer.

Why was this example not allowed by the previous transformer? Was `bar` treated as a local or a global? If it was treated as a local, can simply changing the line 

```python
exec(compile(source, f'<staging:{func.__name__}>', 'exec'), caller_env)
```

to

```
exec(compile(source, f'<staging:{func.__name__}>', 'exec'), caller_env, {})
```

solve the problem?

Reply:
It is assumed to be in the global namespace in this example. In the previous implementation, when`@transform` is called, the global dictionary does not yet contain `bar`, so the combined (thus copied) `caller_env` has no `"bar"` entry in it. Even though it later goes into the global dictionary, the `caller_env` is not possible to track the update. The current implementation directly uses the original globals dictionary and should work.

If in some function's local namespace, it would require us to pass the captured cell instead of `cell_content` into the function and fetch `cell_content` per access; it's not yet implemented.

Reply:
The `local` handling is done and you may review it.